webapp/src/Controller/BaseController.php

```diff
@@ -351,4 +364,61 @@ protected function getDependentEntities(string $entityClass, array $relations)
 
         return $result;
     }
+
+    /**
+     * Get the contests that an event for the given entity should be triggered on
+     * @param                 $entity
+     * @param DOMJudgeService $dj
+     *
+     * @return Contest[]
+     */
+    protected function contestsForEntity($entity, DOMJudgeService $dj) {
+        // Determine contests to emit an event for for the given entity:
+        // * If the entity is a Problem entity, use the getContest()
+        //   of every contest problem in getContestProblems()
+        // * If the entity is a Team (category) entity, get all active
+        //   contests and use those that are open to all teams or the
+        //   team (category) belongs to
+        // * If the entity is a contest, use that
+        // * If the entity has a getContest() method, use that
+        // * If the entity has a getContests() method, use that
+        // Otherwise use the currently active contests
+        $contests = [];
+        if ($entity instanceof Team || $entity instanceof TeamAffiliation) {
+            $possibleContests = $dj->getCurrentContests();
+            foreach ($possibleContests as $contest) {
+                if ($entity->inContest($contest)) {
+                    $contests[] = $contest;
+                }
+            }
+        } elseif ($entity instanceof Problem) {
+            foreach ($entity->getContestProblems() as $contestProblem) {
+                $contests[] = $contestProblem->getContest();
+            }
+        } elseif ($entity instanceof Contest) {
+            $contests = [$entity];
+        } elseif (method_exists($entity, 'getContest')) {
+            $contests = [$entity->getContest()];
+        } elseif (method_exists($entity, 'getContests')) {
+            $contests = $entity->getContests();
+        } else {
+            $contests = $dj->getCurrentContests();
+        }
+
+        return $contests;
+    }
+
+    /**
+     * Return the table name for the given entity
+     * @param $entity
+     *
+     * @return string
+     */
+    protected function tableForEntity($entity)
```

Comment:
Can't this be a static method outside this class? It doesn't depend on anything in this class, right?

Reply:
Any suggestion for where to move it?

Reply:
`webapp/src/Utils/Utils.php`?